service/k8sclient/kubectl/kubectl.go

```diff
@@ -0,0 +1,130 @@
+)
+
+// errNotFound is an error in case of object not found.
+var errNotFound = errors.New("object not found")
+
+// KubeCtl wraps kubectl CLI with version selection and kubeconfig handling.
+type KubeCtl struct {
+	l   logger.Logger
+	cmd []string
+}
+
+// NewKubeCtl creates a new KubeCtl object with a given logger.
+func NewKubeCtl(l logger.Logger) *KubeCtl {
+	// TODO Handle kubectl versions https://jira.percona.com/browse/PMM-6348
+
+	// TODO Handle kubeconfig https://jira.percona.com/browse/PMM-6347
+
+	cmd := []string{"dbaas-kubectl-1.16"}
+	path, err := exec.LookPath(cmd[0])
+	l.Debugf("path: %s, err: %v", path, err)
+	if e, ok := err.(*exec.Error); err != nil && ok && e.Err == exec.ErrNotFound {
+		cmd = []string{"minikube", "kubectl", "--"}
+	}
+	l.Debugf("Using %q", strings.Join(cmd, " "))
+
+	return &KubeCtl{
+		l:   l.WithField("component", "kubectl"),
+		cmd: cmd,
+	}
+}
+
+// Cleanup removes temporary files created by that object.
+func (k *KubeCtl) Cleanup() {
+	// TODO Remove kubeconfig file https://jira.percona.com/browse/PMM-6347
+}
+
+// Get executes `kubectl get` with given object kind and optional name,
+// and decodes resource into `res`.
+// It returns `errNotFound` if object is not found.
+func (k *KubeCtl) Get(ctx context.Context, kind string, name string, res interface{}) error {
+	args := []string{"get", "-o=json", kind}
+	if name != "" {
+		args = append(args, name)
+	}
+
+	stdout, stderr, err := k.run(ctx, args, nil)
+	if err != nil {
+		if strings.Contains(string(stderr), "not found") { // FIXME
+			return errNotFound
+		}
+		return err
+	}
+
+	return json.Unmarshal(stdout, res)
+}
+
+// Apply executes `kubectl apply` with given resource.
+func (k *KubeCtl) Apply(ctx context.Context, res meta.Object) error {
+	_, _, err := k.run(ctx, []string{"apply", "-f", "-"}, res)
+	return err
+}
+
+// Delete executes `kubectl delete` with given resource.
+func (k *KubeCtl) Delete(ctx context.Context, res meta.Object) error {
+	_, _, err := k.run(ctx, []string{"delete", "-f", "-"}, res)
+	return err
+}
+
+// run executes kubectl with given arguments and stdin data (encoded as JSON),
+// and returns stdout, stderr and execution error.
+func (k *KubeCtl) run(ctx context.Context, args []string, stdin interface{}) ([]byte, []byte, error) {
+	args = append(k.cmd, args...)
+
+	var inBuf bytes.Buffer
+	if stdin != nil {
+		e := json.NewEncoder(&inBuf)
+		e.SetIndent("", "  ")
+		if err := e.Encode(stdin); err != nil {
+			return nil, nil, err
+		}
+		k.l.Debugf("Running %s with input:\n%s", strings.Join(args, " "), inBuf.String())
+	} else {
+		k.l.Debugf("Running %s", strings.Join(args, " "))
+	}
+
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...) //nolint:gosec
+	pdeathsig.Set(cmd, unix.SIGKILL)
+	cmd.Stdin = &inBuf
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+	err := cmd.Run()
+
+	k.l.Debugf(outBuf.String()) // FIXME
```

Comment:
That will not work if output contains `%s` or something similar. We can't use XXXf functions as just XXX.

Reply:
Added new Debug method.

Reply:
Why add more methods to interface and implementations instead of using `Debugf("%s", outBuf)`?

Reply:
ok, anything else?

Reply:
Added other methods, since we already have the same usages for Fatal, Info and Debug.